<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Poincaré and SimBio: a versatile and extensible Python ecosystem for modeling systems.</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="article_files/libs/clipboard/clipboard.min.js"></script>
<script src="article_files/libs/quarto-html/quarto.js"></script>
<script src="article_files/libs/quarto-html/popper.min.js"></script>
<script src="article_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="article_files/libs/quarto-html/anchor.min.js"></script>
<link href="article_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="article_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="article_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="article_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="article_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="article.css">
</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Poincaré and SimBio: a versatile and extensible Python ecosystem for modeling systems.</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Mauro Silberberg <a href="https://orcid.org/0000-0002-2402-1100" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Universidad de Buenos Aires, Facultad de Ciencias Exactas y Naturales, Departamento de Física. Buenos Aires, Argentina.
          </p>
        <p class="affiliation">
            CONICET - Universidad de Buenos Aires, Instituto de Física de Buenos Aires (IFIBA). Buenos Aires, Argentina
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Rahuman S. Malik-Sheriff <a href="https://orcid.org/0000-0003-0705-9809" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://ebi.ac.uk/">
            European Bioinformatics Institute, European Molecular Biology Laboratory (EMBL-EBI), Wellcome Trust Genome Campus, Cambridge, UK
            </a>
          </p>
      </div>
      <div class="quarto-title-meta-contents">
    <p class="author">Hernán E. Grecco <a href="https://orcid.org/0000-0002-1165-4320" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Universidad de Buenos Aires, Facultad de Ciencias Exactas y Naturales, Departamento de Física. Buenos Aires, Argentina.
          </p>
        <p class="affiliation">
            CONICET - Universidad de Buenos Aires, Instituto de Física de Buenos Aires (IFIBA). Buenos Aires, Argentina
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Abstract</div>
    <p>Chemical Reaction Networks (CRNs) are a fundamental concept of modelling in numerous fields, including systems biology, biochemistry, chemical engineering and epidemiology. Formulating them in a higher-level description allows to solve them with different approaches, such as deterministic or stochastic simulations. Existing Python tools for CRN modeling wrap external C/C++ libraries for modelling and simulation, which limits their extensibility to new simulation approaches or to reuse packages from the vast Python community. This work introduces two novel Python packages: Poincaré, for definition and simulation of dynamical systems, and SimBio, which builds on top of Poincaré to facilitate the definition of Chemical Reaction Networks (CRNs), and add support for the Systems Biology Markup Language (SBML). Poincaré and SimBio are pure Python packages, which allows to extend their simulation capabilities by writing Python or leveraging other Python packages. But this does not compromise performance, as code can be Just-In-Time compiled with Numba, and benchmarks show comparable or superior performance to other tools. Additionally, to ensure a user-friendly experience, they use of standard typed modern Python syntax that provide a seamless integration with Integrated Development Environments (IDEs), enhances code analysis tools, enabling error identification before execution, and facilitates code completion and refactoring. We believe that this Python-centric approach positions Poincaré and SimBio as valuable tools for the modeling community.</p>
  </div>
</div>

</header>

<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>Chemical Reaction Networks (CRNs) are a fundamental concept of modelling in numerous fields including systems biology, biochemistry, chemical engineering and epidemiology. They are comprised of a set of chemical species or biological entities and representation of complex interaction and transformation between them through a series reactions. These systems can be modelled and simulated through multiple approaches: deterministic Ordinary Differential Equations (ODEs) to model macroscopic behavior, Stochastic Differential Equations (SDEs) to model microscopic fluctuations, and jump processes (Gillespie-like simulations) to account for the discreteness of populations. Instead of directly writing the equations for each of these formulations which is error-prone and difficult to reuse, these models can be written in a higher-level description that can be compiled for these different types of simulations.</p>
<p>Several tools already exist for defining and simulating CRNs. BioSimulators.org <span class="citation" data-cites="shaikhBioSimulatorsCentralRegistry2022">(<a href="#ref-shaikhBioSimulatorsCentralRegistry2022" role="doc-biblioref">Shaikh et al. 2022</a>)</span>, a registry of simulation tools, list at least 15 Python software including COPASI <span class="citation" data-cites="hoopsCOPASICOmplexPAthway2006">(<a href="#ref-hoopsCOPASICOmplexPAthway2006" role="doc-biblioref">Hoops et al. 2006</a>)</span>, Tellurium <span class="citation" data-cites="choiTelluriumExtensiblePythonbased2018">(<a href="#ref-choiTelluriumExtensiblePythonbased2018" role="doc-biblioref">Choi et al. 2018</a>)</span> and PySB <span class="citation" data-cites="lopezProgrammingBiologicalModels2013">(<a href="#ref-lopezProgrammingBiologicalModels2013" role="doc-biblioref">Lopez et al. 2013</a>)</span>. COPASI is a standalone software with a GUI for defining and simulating models. It’s widely used for its user-friendly interface and comprehensive features. There are also python bindings and interfaces for COPASI to allow advanced scripting. Tellurium is a Python-based modeling environment, however, it uses a C++ library called libRoadRunner in the backend to simulate models. PySB is a Python library that created a DSL using standard Python to define models which are then compiled to ODE using a Perl library called BioNetGen <span class="citation" data-cites="harrisBioNetGenAdvancesRulebased2016">(<a href="#ref-harrisBioNetGenAdvancesRulebased2016" role="doc-biblioref">L. A. Harris et al. 2016</a>)</span>.</p>
<p>One limitation of these tools is that they are not extensible from Python, as they are not fully Python packages but wrap libraries in other languages to do the simulation. While they enable model definition and simulation control via Python scripts, they don’t fully leverage Python’s extensive package ecosystem. For example, in COPASI, users are restricted to predefined distributions for Random Parameter Scans and cannot utilize the diverse distributions available in <code>scipy.stats</code>. Similarly, Tellurium doesn’t allow the use of Python solvers, as adding new integrators requires C++ implementation.</p>
<p>Another challenge is the way models are written. Many tools use a Domain Specific Language (DSL) for coding, and support the System Biology Markup Language (SBML) as an exchange format <span class="citation" data-cites="huckaSBMLL3V2">(<a href="#ref-huckaSBMLL3V2" role="doc-biblioref">Hucka et al., n.d.</a>)</span>, as direct SBML coding is impractical. A DSL allows reuse of code in different programming environments, but it will not be recognized by default in Integrated Development Environments (IDEs). Therefore DSLs cannot provide the development experience such as syntax highlighting, code completion, refactoring, and static analysis, unless such support is specifically developed. Tellurium uses a Domain Specific Language (DSL) called Antimony <span class="citation" data-cites="smithAntimonyModularModel2009">(<a href="#ref-smithAntimonyModularModel2009" role="doc-biblioref">Smith et al. 2009</a>)</span>, which can be translated to SBML. An extension for Visual Studio Code was developed, but its maintenance could be demanding task for the systems biology community. PySB, using Python’s dynamic nature, created a DSL within Python. By default, it uses global state to create species and parameters, without assigning them to Python variables or adding them explicitly to the model, but this approach is not fully compatible with IDEs, affecting the development experience.</p>
<p>To overcome these limitations, we developed poincaré and SimBio, open-source Python packages for defining and simulating systems. Poincaré allows defining differential equation systems, while SimBio builds on it for defining reaction networks. They are focused on providing an ergonomic experience to end-users by integrating well with IDEs and static analysis tools through the use of standard modern Python syntax. Since they are coded in Python, every part from model definition to simulation can be extended from Python syntax. Being the first-ever pure Python packages for systems modelling, they offer extensive extensibility, from simple tasks like reusing integrators defined in other packages, to complex ones like altering the compilation process to leverage some structure in the equations. For example, using a for-loop in the compiled equations could improve the runtime performance if there is some repetitive structure in the system, as happens in spatial modelling. The models built using these packages can be introspected to create other representations, such as graphs connecting species and/or reactions, or tables with parameters or equations. Furthermore, they have a modular architecture with a clear separation of concerns, making it easier to maintain or to contribute new code, which is beneficial for developers and maintainers. We showcased the reliability of these tools by benchmarking them against the simulation results from other tools. We also highlighted the substantial performance improvements our tools offer, as this is crucial for construction and simulation of models of whole cells and organisms, which necessitate the simulation of significantly large-scale models.</p>
</section>
<section id="results" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Results</h1>
<p>Modular code architecture makes code reusable, extensible, and easier to maintain. Therefore, we split the code to define and simulate reaction systems into three Python packages: symbolite, to create symbolic expressions; poincaré, to define dynamical systems; and simbio, to define reaction systems and interface with systems biology standards such as SBML. These are pure Python packages with standard dependencies from the PyData scientific stack such as NumPy <span class="citation" data-cites="harrisArrayProgrammingNumPy2020">(<a href="#ref-harrisArrayProgrammingNumPy2020" role="doc-biblioref">C. R. Harris et al. 2020</a>)</span> and pandas <span class="citation" data-cites="mckinneyDataStructuresStatistical2010">(<a href="#ref-mckinneyDataStructuresStatistical2010" role="doc-biblioref">McKinney 2010</a>)</span>. They are published in PyPI (the Python Package Index), where links to the source code and documentation hosted in GitHub can be found, and can be easily installed with <code>pip install &lt;package_name&gt;</code>.</p>
<p>Symbolite is a lightweight symbolics package to create algebraic mathematical expressions. Symbolite expressions can be inspected and compiled to various backends. Currently, we have implementations for NumPy <span class="citation" data-cites="harrisArrayProgrammingNumPy2020">(<a href="#ref-harrisArrayProgrammingNumPy2020" role="doc-biblioref">C. R. Harris et al. 2020</a>)</span>; Numba <span class="citation" data-cites="lamNumbaLLVMbasedPython2015">(<a href="#ref-lamNumbaLLVMbasedPython2015" role="doc-biblioref">Lam, Pitrou, and Seibert 2015</a>)</span>, a Just-in-Time (JIT) compiler to LLVM; SymPy <span class="citation" data-cites="meurerSymPySymbolicComputing2017">(<a href="#ref-meurerSymPySymbolicComputing2017" role="doc-biblioref">Meurer et al. 2017</a>)</span>, a library for symbolic mathematics; and JAX <span class="citation" data-cites="jax2018github">(<a href="#ref-jax2018github" role="doc-biblioref">Bradbury et al. 2018</a>)</span>, a library that support automatic differentiation and compilation to GPUs and TPUs. Symbolite is designed to facilitate the easy integration of new backends.</p>
<section id="versatile-modelling-and-simulation-of-dynamical-systems-with-poincaré" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="versatile-modelling-and-simulation-of-dynamical-systems-with-poincaré"><span class="header-section-number">2.1</span> Versatile modelling and simulation of dynamical systems with Poincaré</h2>
<div id="fig-poincare" class="quarto-layout-panel">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell quarto-layout-cell" data-execution_count="2" style="flex-basis: 100.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> poincare <span class="im">import</span> Derivative, System, Variable, initial</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-first-order" class="quarto-layout-panel quarto-layout-cell quarto-layout-cell-subref" data-ref-parent="fig-poincare" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="cell quarto-layout-cell" data-execution_count="3" style="flex-basis: 33.3%;justify-content: center;">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Decay(System):</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    x: Variable <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    eq <span class="op">=</span> x.derive() <span class="op">&lt;&lt;</span> <span class="op">-</span>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
\frac{dx}{dt} = -x
\]</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
x(0) = 1
\]</span></p>
</div>
</div>
<p></p><figcaption class="figure-caption">(a) First-order system of an exponential decay. The variable <code>x</code> stores the initial condition for <span class="math inline">\(x\)</span>, and the variable <code>eq</code> stores the rate equation for <span class="math inline">\(x\)</span>.</figcaption><p></p>
</figure>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-second-order" class="quarto-layout-panel quarto-layout-cell quarto-layout-cell-subref" data-ref-parent="fig-poincare" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="cell quarto-layout-cell" data-execution_count="4" style="flex-basis: 33.3%;justify-content: center;">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Oscillator(System):</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    x: Variable <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    v: Derivative <span class="op">=</span> x.derive(initial<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    eq <span class="op">=</span> v.derive() <span class="op">&lt;&lt;</span> <span class="op">-</span>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
\frac{d^2x}{dt^2} = -x
\]</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
\begin{cases}
    x(0) &amp;= 1 \\
    \frac{dx}{dt}(0) &amp;= 0
\end{cases}
\]</span></p>
</div>
</div>
<p></p><figcaption class="figure-caption">(b) Second-order system of an harmonic oscillator, where the variable <code>v</code> stores the derivative <span class="math inline">\(\frac{dx}{dt}\)</span> and the rate equation is specified for the derivative <code>v</code>.</figcaption><p></p>
</figure>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-composition" class="quarto-layout-panel quarto-layout-cell quarto-layout-cell-subref" data-ref-parent="fig-poincare" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="cell quarto-layout-cell" data-execution_count="5" style="flex-basis: 33.3%;justify-content: center;">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> BigModel(System):</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    x: Variable <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    linked <span class="op">=</span> Decay(x<span class="op">=</span>x)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    independent <span class="op">=</span> Decay(x<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
\begin{cases}
    \frac{dx}{dt} = -x \\
    \frac{dy}{dt} = -y
\end{cases}
\]</span></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 33.3%;justify-content: center;">
<p><span class="math display">\[
\begin{cases}
    x(0) &amp;= 1 \\
    y(0) &amp;= 2
\end{cases}
\]</span></p>
</div>
</div>
<p></p><figcaption class="figure-caption">(c) First-order system of two exponential decays by composition of the <code>Decay</code> system. The subsystem <code>linked</code> has a reference to the outer variable <code>x</code>, while the subsystem <code>independent</code> defines a new variable with initial condition §2$, which on the corresponding mathematical expression was named <span class="math inline">\(y\)</span>.</figcaption><p></p>
</figure>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Code and corresponding mathematical expressions for different systems.</figcaption><p></p>
</figure>
</div>
<p>Poincare is a package to define and simulate dynamical systems. It provides a <code>System</code> class, where one can define <code>Constant</code>s, <code>Parameter</code>s, <code>Variable</code>s, and create equations linking a variable’s derivative with an expression (<a href="#fig-first-order">Figure&nbsp;1 (a)</a>). It also allows to define higher-order systems by assigning an initial condition to a <code>Derivative</code> (<a href="#fig-second-order">Figure&nbsp;1 (b)</a>).</p>
<p>Utilizing classes for system definition offers several advantages:</p>
<ol type="1">
<li>The variable name to which a component is assigned can be automatically saved in the component for introspection (i.e., <code>Oscilator.x.name == "x"</code>),</li>
<li>It provides a namespace such that allows to easily define of multiple independent models in the same script,</li>
<li>It allows IDEs to provide autocomplete and refactoring capabilities (<code>Oscillator.&lt;TAB&gt;</code> shows <code>x</code>, <code>v</code> and <code>eq</code>),</li>
<li>It allows creation of instances which can be composed into a bigger model (<a href="#fig-composition">Figure&nbsp;1 (c)</a>).</li>
</ol>
<p>For this last point, IDEs that support <code>dataclass_transform</code> <span class="citation" data-cites="debontePEP681Data2021">(<a href="#ref-debontePEP681Data2021" role="doc-biblioref">De Bonte and Traut 2021</a>)</span> can provide a tooltip with the expected signature (<a href="#fig-composition">Figure&nbsp;1 (c)</a>). This requires the use of type annotations which play a more significant role in static type checking, as they can help to identify errors before running the code. For instance, to parameterize the initial conditions of variables we have to use a <code>Constant</code>. If we try to use a <code>Parameter</code>, which could be a time-dependent expression, it is flagged as a type error (<a href="#fig-ide">Figure&nbsp;2</a>).</p>
<div id="fig-ide" class="quarto-layout-panel">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-bottom">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="src/ide/ide1.svg" class="img-fluid figure-img"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="src/ide/ide2.svg" class="img-fluid figure-img"></p>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Screenshots of Visual Studio Code showing tooltips and highlighted type errors. <code>Constant</code> are assigned with <code>assign(..., constant=True)</code> and can be used to link <code>Variable</code>s initial conditions, while trying to use a <code>Parameter</code> instead is flagged as type errors (red underlining). The IDE automatically recognizes <code>e</code> as an <code>Equation</code>, and provides autocompletion of the variables. A tooltip is shown when composing models, which show the expected variables and their default values. The IDE highlights wrong names (<code>z</code> is not a name in <code>Model</code>) and mismatched types (<code>x</code> is <code>Variable</code> and <code>a</code> must be a number or a <code>Constant</code>)</figcaption><p></p>
</figure>
</div>
<p>To simulate a system, we created a <code>Simulator</code> instance (<a href="#fig-sim">Figure&nbsp;3</a>), which compiles a given system and interfaces with solvers. By default, it creates a first-order ordinary differential equation (ODE) system using <code>numpy</code> as a backend. This can be easily switched to other solvers. The Simulator wraps the output in a <code>pandas.DataFrame</code>, which can be easily plotted with the standard <code>plot</code> method.</p>
<div id="fig-sim" class="quarto-layout-panel">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell quarto-layout-cell" data-execution_count="6" style="flex-basis: 100.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> poincare <span class="im">import</span> Simulator</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>sim <span class="op">=</span> Simulator(Oscillator)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> sim.solve(save_at<span class="op">=</span>np.linspace(<span class="dv">0</span>, <span class="dv">5</span>, <span class="dv">101</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell quarto-layout-cell" data-execution_count="7" style="flex-basis: 50.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">v</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">time</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0.00</td>
<td>1.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0.05</td>
<td>0.998750</td>
<td>-0.049979</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0.10</td>
<td>0.995005</td>
<td>-0.099835</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0.15</td>
<td>0.988772</td>
<td>-0.149440</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0.20</td>
<td>0.980069</td>
<td>-0.198672</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell quarto-layout-cell" data-execution_count="8" style="flex-basis: 50.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>df.plot()<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="article_files/figure-html/cell-9-output-1.png" width="358" height="208" class="figure-img"></p>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;3: Simulation of the <code>Oscillator</code> system from <a href="#fig-second-order">Figure&nbsp;1 (b)</a>. The output is a <code>pandas.DataFrame</code> with a column for each variable and the time as index. It is inspected and plotted with the <code>pandas</code> methods.</figcaption><p></p>
</figure>
</div>
<p>Switching backends to <code>"numba"</code> can improve model runtime on a factor up to x30 or more for big or long running models, however incurs in a compile time penalty for the first run which must be taken into account.</p>
</section>
<section id="extensible-definition-of-reaction-networks-using-simbio" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="extensible-definition-of-reaction-networks-using-simbio"><span class="header-section-number">2.2</span> Extensible definition of reaction networks using SimBio</h2>
<div id="fig-simbio" class="quarto-layout-panel">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell quarto-layout-cell" data-execution_count="9" style="flex-basis: 100.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> simbio <span class="im">import</span> Compartment, MassAction, Species, Reaction, initial</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-panel quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="cell quarto-layout-cell" data-execution_count="10" style="flex-basis: 50.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Model(Compartment):</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""2A -&gt; B"""</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    A: Species <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    B: Species <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    r <span class="op">=</span> Reaction(</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>        reactants<span class="op">=</span>[<span class="dv">2</span> <span class="op">*</span> A],</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>        products<span class="op">=</span>[B],</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>        rate_law<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><span class="math display">\[
\begin{align}
&amp;\begin{cases}
    \frac{dA}{dt} = -2 \\
    \frac{dB}{dt} = +1
\end{cases}
\\\\
&amp;\begin{cases}
    A(0) &amp;= 1 \\
    B(0) &amp;= 0
\end{cases}
\end{align}
\]</span></p>
</div>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="quarto-layout-panel quarto-layout-cell" style="flex-basis: 100.0%;justify-content: center;">
<div class="quarto-layout-row quarto-layout-valign-center">
<div class="cell quarto-layout-cell" data-execution_count="11" style="flex-basis: 50.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Model(Compartment):</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""2A -&gt; B"""</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    A: Species <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    B: Species <span class="op">=</span> initial(default<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    r <span class="op">=</span> MassAction(</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        reactants<span class="op">=</span>[<span class="dv">2</span> <span class="op">*</span> A],</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>        products<span class="op">=</span>[B],</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>        rate<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><span class="math display">\[
\begin{align}
&amp;\begin{cases}
    \frac{dA}{dt} = -2 A^2 \\
    \frac{dB}{dt} = +1 A^2
\end{cases}
\\\\
&amp;\begin{cases}
    A(0) &amp;= 1 \\
    B(0) &amp;= 0
\end{cases}
\end{align}
\]</span></p>
</div>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;4: A reaction system for species <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> with initial conditions <span class="math inline">\(1\)</span> and <span class="math inline">\(0\)</span>, respectively. A single reaction transforming <span class="math inline">\(2A\)</span> into <span class="math inline">\(B\)</span> is saved in variable <code>r</code>. The rate <span class="math inline">\(1\)</span> is specified directly for <code>Reaction</code>, and is proportional to the reactants for <code>MassAction</code>.</figcaption><p></p>
</figure>
</div>
<p>For the reaction networks, our focus is on first-order differential equations that describe the rate of change of species. SimBio simplifies the definition of these network models by introducing <code>Species</code>, which consists of a <code>poincare.Variable</code> and a stoichiometric number, and <code>Reaction</code>, a construct that converts reactants into products taking into account the stoichiometry (<a href="#fig-simbio">Figure&nbsp;4</a>). Additionally, SimBio features <code>MassAction</code>, a subclass of <code>Reaction</code>, which intuitively incorporates reactants into the rate law (<a href="#fig-simbio">Figure&nbsp;4</a>).</p>
<p>Several commonly used reactions are predefined as <code>MassAction</code> subclasses, such as <code>MichaelisMenten</code> (<span class="math inline">\(S + E \leftrightarrow ES \rightarrow P + E\)</span>) and its approximate form without the intermediate species <span class="math inline">\(ES\)</span>, and it is also simple to implement your own as subclasses of <code>Reaction</code> or <code>MassAction</code>. Additionally, SimBio supports importing from and exporting to SBML, and downloading them directly from BioModels <span class="citation" data-cites="malik-sheriffBioModels15Years2020">(<a href="#ref-malik-sheriffBioModels15Years2020" role="doc-biblioref">Malik-Sheriff et al. 2020</a>)</span> (<a href="#fig-simbio-io">Figure&nbsp;5</a>).</p>
<div id="fig-simbio-io" class="quarto-layout-panel">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell quarto-layout-cell" data-execution_count="13" style="flex-basis: 100.0%;justify-content: center;">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> simbio.io <span class="im">import</span> biomodels, sbml</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> sbml.load(<span class="st">"repressilator.sbml"</span>)  <span class="co"># from existing SBML file</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> biomodels.load_model(<span class="st">"BIOMD12"</span>)  <span class="co"># from BioModels</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">

<table>
<thead>
</thead><colgroup span="3"></colgroup><tbody><tr><th colspan="3" scope="colgroup">Elowitz2000 - Repressilator</th></tr>
<tr><th>type      </th><th style="text-align: right;">  total</th><th>names                                                         </th></tr>

</tbody><tbody>
<tr><td>variables </td><td style="text-align: right;">      6</td><td>PX, PY, PZ, X, Y, Z                                           </td></tr>
<tr><td>parameters</td><td style="text-align: right;">     17</td><td>cell, beta, alpha0, alpha, eff, n, KM, tau_mRNA, tau_prot, ...</td></tr>
<tr><td>equations </td><td style="text-align: right;">     12</td><td>Reaction1, Reaction2, Reaction3, Reaction4, Reaction5, ...    </td></tr>
</tbody>
</table>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;5: Creation of a model from a local SBML file or one uploaded to BioModels.</figcaption><p></p>
</figure>
</div>
</section>
<section id="reproducibility-and-performance" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="reproducibility-and-performance"><span class="header-section-number">2.3</span> Reproducibility and performance</h2>
<div id="fig-runtime" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="src/performance/figures/performance.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;6: Performance of different softwares to solve models from the curated section of BioModels. (left) Run time for the model BIOMD3 as a function of the number of output points. (right) Run time for different models for 300 output points, using the geometric mean of the different softwares to order them.</figcaption>
</figure>
</div>
<p>To evaluate SimBio’s reproducibility, we analyzed curated SBML models from BioModels <span class="citation" data-cites="malik-sheriffBioModels15Years2020">(<a href="#ref-malik-sheriffBioModels15Years2020" role="doc-biblioref">Malik-Sheriff et al. 2020</a>)</span>. Among the first 100 curated models from BioModels we selected 60 which did not contains events, as SimBio doesn’t support events yet. We simulated the selected models with COPASI and used the simulated results as ground truth, and demonstrated that SimBio reproduces the results with minor numerical differences depending on the solver tolerance.</p>
<p>For performance testing, we ran simulations using COPASI, Tellurium/RoadRunner, and SimBio. Within SimBio, both NumPy and Numba backends were considered. The LSODA solver was used for COPASI and SimBio, while for Tellurium the comparable CVODE solver was used. In all cases, we used relative and absolute tolerances of <span class="math inline">\(10^{-6}\)</span>. We measured three simulation stages: loading, the initial (cold) run, and subsequent (warm) runs for each model.</p>
<p>For COPASI and Tellurium/RoadRunner, we noted that its runtime depended on the number of evaluation points, something that does not seem to happen with SimBio (<a href="#fig-runtime">Figure&nbsp;6</a>, left). While SimBio’s NumPy backend is slower than both COPASI and RoadRunner, we obtained an order of magnitude speed-up using the numba backend putting it on par with them. A user might have to consider the trade-off between compilation and run times, as the compilation of the right-hand-side (RHS) code might take longer than the runtime itself, and not be worth it for running the model only once. Another speed-up in the runtime can be had by switching the LSODA <code>scipy</code> solver for a more efficient <code>numbalsoda</code> implementation, which avoids the Python interpreter between each of the integration steps. This last combination beats all other methods, which is also true for the other models we tested (<a href="#fig-runtime">Figure&nbsp;6</a>, right).</p>
</section>
</section>
<section id="discussion" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Discussion</h1>
<p>In this article, we introduced a suite of Python packages we developed for defining and simulating dynamical systems and reaction networks. These packages are deeply integrated with Integrated Development Environments (IDEs), enabling code analysis tools to identify errors prior to execution and assist in refactoring and code completion. We adopted standard modern Python syntax to ensure seamless IDE integration, supported by the extensive Python community.</p>
<p>Our approach differs from previous tools in that both the model definition and its compilation into an Ordinary Differential Equation (ODE) function are entirely Python-based. This approach simplifies the development of various simulation methods, including performance enhancements that exploit specific model structures. Importantly, being Python-based does not compromise performance compared to C/C++ tools, as the ODE functions can be Just-In-Time (JIT) compiled using Numba.</p>
<p>The inclusion of SBML support facilitates the effortless reuse of models created by the systems biology community, along with the vast collection of public models hosted in the BioModels repository. The modular architecture of these packages facilitates their reuse, enhancement, and extension by the wider Python community. For instance, an individual from outside the systems biology field could contribute a stochastic integrator to poincaré, which would then be available in SimBio. This clear separation of concerns also makes the packages more comprehensible, lowering the barrier for contributing improvements or new features. Such an architecture ensures their maintainability and ongoing development well into the future.</p>
</section>
<section id="references" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-jax2018github" class="csl-entry" role="listitem">
Bradbury, James, Roy Frostig, Peter Hawkins, Matthew James Johnson, Chris Leary, Dougal Maclaurin, George Necula, et al. 2018. <span>“<span>JAX</span>: Composable Transformations of <span>Python</span>+<span>NumPy</span> Programs.”</span>
</div>
<div id="ref-choiTelluriumExtensiblePythonbased2018" class="csl-entry" role="listitem">
Choi, Kiri, J Kyle Medley, Matthias König, Kaylene Stocking, Lucian Smith, Stanley Gu, and Herbert M Sauro. 2018. <span>“Tellurium: <span>An</span> Extensible Python-Based Modeling Environment for Systems and Synthetic Biology.”</span> <em>Bio Systems</em> 171 (September): 74–79. <a href="https://doi.org/10.1016/j.biosystems.2018.07.006">https://doi.org/10.1016/j.biosystems.2018.07.006</a>.
</div>
<div id="ref-debontePEP681Data2021" class="csl-entry" role="listitem">
De Bonte, Erik, and Eric Traut. 2021. <span>“<span>PEP</span> 681 <span></span> <span>Data Class Transforms</span>.”</span> <em>Python Enhancement Proposals</em>. https://peps.python.org/pep-0681/.
</div>
<div id="ref-harrisArrayProgrammingNumPy2020" class="csl-entry" role="listitem">
Harris, Charles R., K. Jarrod Millman, Stéfan J. van der Walt, Ralf Gommers, Pauli Virtanen, David Cournapeau, Eric Wieser, et al. 2020. <span>“Array Programming with <span>NumPy</span>.”</span> <em>Nature</em> 585 (7825): 357–62. <a href="https://doi.org/10.1038/s41586-020-2649-2">https://doi.org/10.1038/s41586-020-2649-2</a>.
</div>
<div id="ref-harrisBioNetGenAdvancesRulebased2016" class="csl-entry" role="listitem">
Harris, Leonard A, Justin S Hogg, José-Juan Tapia, John A P Sekar, Sanjana Gupta, Ilya Korsunsky, Arshi Arora, Dipak Barua, Robert P Sheehan, and James R Faeder. 2016. <span>“<span>BioNetGen</span> 2.2: Advances in Rule-Based Modeling.”</span> <em>Bioinformatics (Oxford, England)</em> 32 (21): 3366–68. <a href="https://doi.org/10.1093/bioinformatics/btw469">https://doi.org/10.1093/bioinformatics/btw469</a>.
</div>
<div id="ref-hoopsCOPASICOmplexPAthway2006" class="csl-entry" role="listitem">
Hoops, Stefan, Sven Sahle, Ralph Gauges, Christine Lee, Jürgen Pahle, Natalia Simus, Mudita Singhal, Liang Xu, Pedro Mendes, and Ursula Kummer. 2006. <span>“<span>COPASI</span><span></span>a <span>COmplex PAthway SImulator</span>.”</span> <em>Bioinformatics</em> 22 (24): 3067–74. <a href="https://doi.org/10.1093/bioinformatics/btl485">https://doi.org/10.1093/bioinformatics/btl485</a>.
</div>
<div id="ref-huckaSBMLL3V2" class="csl-entry" role="listitem">
Hucka, Michael, Frank Bergmann, Stefan Hoops, Sarah M Keating, Nicolas Le Novère, Chris J Myers, Brett G Olivier, et al. n.d. <span>“<span>SBML L3V2</span>.”</span>
</div>
<div id="ref-lamNumbaLLVMbasedPython2015" class="csl-entry" role="listitem">
Lam, Siu Kwan, Antoine Pitrou, and Stanley Seibert. 2015. <span>“Numba: A <span class="nocase">LLVM-based Python JIT</span> Compiler.”</span> In <em>Proceedings of the <span>Second Workshop</span> on the <span>LLVM Compiler Infrastructure</span> in <span>HPC</span></em>, 1–6. <span>LLVM</span> ’15. <span>New York, NY, USA</span>: <span>Association for Computing Machinery</span>. <a href="https://doi.org/10.1145/2833157.2833162">https://doi.org/10.1145/2833157.2833162</a>.
</div>
<div id="ref-lopezProgrammingBiologicalModels2013" class="csl-entry" role="listitem">
Lopez, Carlos F, Jeremy L Muhlich, John A Bachman, and Peter K Sorger. 2013. <span>“Programming Biological Models in <span>Python</span> Using <span>PySB</span>.”</span> <em>Molecular Systems Biology</em> 9 (January): 646. <a href="https://doi.org/10.1038/msb.2013.1">https://doi.org/10.1038/msb.2013.1</a>.
</div>
<div id="ref-malik-sheriffBioModels15Years2020" class="csl-entry" role="listitem">
Malik-Sheriff, Rahuman S, Mihai Glont, Tung V N Nguyen, Krishna Tiwari, Matthew G Roberts, Ashley Xavier, Manh T Vu, et al. 2020. <span>“<span>BioModels</span><span></span>15 Years of Sharing Computational Models in Life Science.”</span> <em>Nucleic Acids Research</em> 48 (D1): D407–15. <a href="https://doi.org/10.1093/nar/gkz1055">https://doi.org/10.1093/nar/gkz1055</a>.
</div>
<div id="ref-mckinneyDataStructuresStatistical2010" class="csl-entry" role="listitem">
McKinney, Wes. 2010. <span>“Data <span>Structures</span> for <span>Statistical Computing</span> in <span>Python</span>.”</span> In <em>Python in <span>Science Conference</span></em>, 56–61. <span>Austin, Texas</span>. <a href="https://doi.org/10.25080/Majora-92bf1922-00a">https://doi.org/10.25080/Majora-92bf1922-00a</a>.
</div>
<div id="ref-meurerSymPySymbolicComputing2017" class="csl-entry" role="listitem">
Meurer, Aaron, Christopher P. Smith, Mateusz Paprocki, Ondřej Čertík, Sergey B. Kirpichev, Matthew Rocklin, AmiT Kumar, et al. 2017. <span>“<span>SymPy</span>: Symbolic Computing in <span>Python</span>.”</span> <em>PeerJ Computer Science</em> 3 (January): e103. <a href="https://doi.org/10.7717/peerj-cs.103">https://doi.org/10.7717/peerj-cs.103</a>.
</div>
<div id="ref-shaikhBioSimulatorsCentralRegistry2022" class="csl-entry" role="listitem">
Shaikh, Bilal, Lucian P Smith, Dan Vasilescu, Gnaneswara Marupilla, Michael Wilson, Eran Agmon, Henry Agnew, et al. 2022. <span>“<span>BioSimulators</span>: A Central Registry of Simulation Engines and Services for Recommending Specific Tools.”</span> <em>Nucleic Acids Research</em> 50 (W1): W108–14. <a href="https://doi.org/10.1093/nar/gkac331">https://doi.org/10.1093/nar/gkac331</a>.
</div>
<div id="ref-smithAntimonyModularModel2009" class="csl-entry" role="listitem">
Smith, Lucian P., Frank T. Bergmann, Deepak Chandran, and Herbert M. Sauro. 2009. <span>“Antimony: A Modular Model Definition Language.”</span> <em>Bioinformatics</em> 25 (18): 2452–54. <a href="https://doi.org/10.1093/bioinformatics/btp401">https://doi.org/10.1093/bioinformatics/btp401</a>.
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>